In [1]:
import pandas as pd
dataset=pd.read_csv("E:/LetsUpgrade/Day_24/train.csv")
from sklearn import preprocessing
from sklearn.model_selection import train_test_split 
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn import neighbors


## Preprocessing

In [7]:
le=preprocessing.LabelEncoder()
le.fit(dataset["Sex"])
dataset["Sex"]=le.transform(dataset["Sex"])
le.fit(dataset["Embarked"])
dataset["Embarked"]=le.transform(dataset["Embarked"])

## Splitting Dataset

In [8]:
y=dataset["Pclass"]
X=dataset.drop(["Survived","PassengerId","Name","Ticket","Cabin"],axis=1)
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=0)

## Applying KNN 

In [12]:
def funcKNN(k):
    knn=neighbors.KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train,y_train).score(X_test,y_test)
    y_pred=knn.predict(X_test)
    print("\nConfusion matrix when k = ",k,":\n",confusion_matrix(y_test,y_pred))
    print("Accuracy is when k = ",k,":",accuracy_score(y_test,y_pred,normalize=True))
    
for i in range(1,268):
    funcKNN(i)


Confusion matrix when k =  1 :
 [[ 63   5   2]
 [  3  38   8]
 [  0   8 140]]
Accuracy is when k =  1 : 0.9026217228464419

Confusion matrix when k =  2 :
 [[ 67   2   1]
 [  7  39   3]
 [  3  12 133]]
Accuracy is when k =  2 : 0.8951310861423221

Confusion matrix when k =  3 :
 [[ 63   5   2]
 [  7  28  14]
 [  2   3 143]]
Accuracy is when k =  3 : 0.8764044943820225

Confusion matrix when k =  4 :
 [[ 63   5   2]
 [  7  33   9]
 [  3  11 134]]
Accuracy is when k =  4 : 0.8614232209737828

Confusion matrix when k =  5 :
 [[ 61   7   2]
 [  7  30  12]
 [  3   3 142]]
Accuracy is when k =  5 : 0.8726591760299626

Confusion matrix when k =  6 :
 [[ 66   3   1]
 [  7  35   7]
 [  3  10 135]]
Accuracy is when k =  6 : 0.8838951310861424

Confusion matrix when k =  7 :
 [[ 62   7   1]
 [  7  30  12]
 [  3   4 141]]
Accuracy is when k =  7 : 0.8726591760299626

Confusion matrix when k =  8 :
 [[ 62   7   1]
 [  7  35   7]
 [  3  10 135]]
Accuracy is when k =  8 : 0.8689138576779026

Confusi

## Result

**Inference :**
* The accuracy when k=1 is 90.26% and accuracy when k=267, the accuracy is 71.91%
* The accuracy score of the model decreases as the K value increases.
* The accuracy is maximum when k=1
* Hence, the best fitting k value is 1

